cmd/pmm-managed-starlark/main_test.go

```diff
@@ -0,0 +1,122 @@
+// pmm-managed
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"os/exec"
+	"testing"
+
+	"github.com/percona/pmm/api/agentpb"
+	"github.com/stretchr/testify/require"
+
+	"github.com/percona/pmm-managed/services/checks"
+)
+
+var (
+	validQueryActionResult = []map[string]interface{}{
+		{"Value": "5.7.30-33-log", "Variable_name": "version"},
+		{"Value": "Percona Server (GPL), Release 33, Revision 6517692", "Variable_name": "version_comment"},
+		{"Value": "x86_64", "Variable_name": "version_compile_machine"},
+		{"Value": "Linux", "Variable_name": "version_compile_os"},
+		{"Value": "-log", "Variable_name": "version_suffix"},
+	}
+
+	invalidQueryActionResult = []map[string]interface{}{
+		{"key-1": "val-1", "key-2": "val-2"},
+	}
+)
+
+func TestRunChecks(t *testing.T) {
+	testCases := []struct {
+		err     bool
+		version uint32
+		name    string
+		script  string
+		result  []map[string]interface{}
+	}{
+		{
+			err:     true,
+			version: 5,
+			name:    "invalid version",
+			script:  "def check(): return []",
+			result:  invalidQueryActionResult,
+		},
+		{
+			err:     true,
+			version: 1,
+			name:    "invalid starlark syntax",
+			script:  "@ + @",
+			result:  invalidQueryActionResult,
+		},
+		{
+			err:     true,
+			version: 1,
+			name:    "invalid query action result",
+			script:  "def check(): return []",
+			result:  invalidQueryActionResult,
+		},
+		{
+			err:     true,
```

Comment:
I think we should check stderr content via ExitError.Stderr

Reply:
i logged `string(ExitError.Stderr)` in the tests but it turned out to be empty

Reply:
whereas if I log it from `checks.go` I can see the error and the stack trace

Reply:
do you still want me to check it?

Reply:
Yeah, we should ensure that this bad script is killed by our limit, not by generic Linux OOM killer, for example. There should be a way to do that

Reply:
ok I logged the `stderr` of the child process and got the whole stacktrace, and tried to match the first line for each test case, this is what I get from the memory consuming starlark script `fatal error: runtime: out of memory`

I also tried writing a test for CPU time I couldn't find a way to write an infinite loop as starlark uses `for-range` and doesn't allow `while`so I wrote a very long loop instead 

```
def check(rows):
						for x in range(10000000):
							pass
```

which gives me this error `fatal error: runtime: cannot allocate memory` and sometimes I get this `signal killed` which makes it unreliable

I think the reason for that is 

```
RLIMIT_CPU
    CPU time limit in seconds. When the process reaches the soft limit, it is sent a SIGXCPU signal. The default action for this signal is to terminate the process. However, the signal can be caught, and the handler can return control to the main program. If the process continues to consume CPU time, it will be sent SIGXCPU once per second until the hard limit is reached, at which time it is sent SIGKILL. (This latter point describes Linux behavior. Implementations vary in how they treat processes which continue to consume CPU time after reaching the soft limit. Portable applications that need to catch this signal should perform an orderly termination upon first receipt of SIGXCPU.) 
```

so I guess I will try to catch this signal now and kill the process forcefully from the code instead of relying on the OS

Reply:
@AlekSi is there a way to write an infinite loop in starlark? I looked but couldn't find it

Reply:
https://godoc.org/go.starlark.net/resolve, `resolve.AllowRecursion = true` if some special flag is passed to pmm-managed-starlark (that should be passed by default, but will be passed in test)